rfd/0056-teleport-terminal.md

```diff
@@ -0,0 +1,118 @@
+---
+authors: Alexey Kontsevoy (alexey@goteleport.com)
+state: draft
+---
+
+# RFD 0056 - Teleport Terminal
+
+## What
+Teleport Terminal is a desktop application that provides quick access to remote resources via Teleport.
+This RFD defines the high-level architecture of Teleport Terminal.
+
+## Details
+There are two main components to Teleport Terminal:
+1. teleterm
+2. tsh (daemon mode)
+
+### teleterm
+`teleterm` is an [Electron](https://www.electronjs.org/) application that uses Chromium engine
+for its tabbed UI and nodejs for OS-level operations.
+Electron has been chosen because of cross-platform support and an ability to reuse most of existing Teleport Web UI components and design system.
+
+#### UI
+`teleterm` UI will have the following features:
+1. Built-in fully featured terminal based on [xterm](https://xtermjs.org/) and [node-pty](https://github.com/microsoft/node-pty) to fork processes.
+2. Tabbed layout where a tab can be an ssh session, rdp connection, or any other document.
+3. Ability to add and access multiple Teleport clusters at the same time.
+4. Command Palette for quick access and navigation.
+
+### tsh daemon
+`tsh daemon` is a `tsh` tool that runs as a service. A hidden flag launches `tsh` in a background process that exposes gRPC API over unix-socket (similarly to docker daemon).
```

Comment:
We're especially sensitive to `tsh` latency, so let's make sure that all this extra code we add to `tsh` doesn't have a material impact on startup time or other common (non-teleterm) operations.

Reply:
How would we go about doing this? Is there a benchmark suite I could run to test this? I see some stuff in `/examples/bench` and `/lib/benchmark`, but I'm not sure if there's something more suited for what we want to achieve here.

As far as I understood what Alexey told me during onboarding, `tsh daemon start` is the only entrypoint through which the Teleterm code can be executed, so I believe it shouldn't have impact on startup time for other commands. But yeah if we have some benchmark suites available then I definitely want to run them.

https://github.com/gravitational/teleport/blob/c681dd5896e388883ac3afbde394800f2d6ad91e/tool/tsh/tsh.go#L772-L773

https://github.com/gravitational/teleport/blob/c681dd5896e388883ac3afbde394800f2d6ad91e/tool/tsh/daemon.go#L30-L47

Reply:
I don't see this being a major issue (at least not yet). I just wanted to raise it early so that we keep thinking about it as we expand the amount of `tsh` code needed for terminal to function. It just seemed like an interesting design choice to ship extra code to *all users* in `tsh` that is only needed for the terminal use case.

Out of curiosity, what's the size of the `tsh` binary before and after adding the teleterm GRPC server?

Reply:
* `65.062MiB` on `master` (b99613507, the base of this branch at the moment)
* `65.577MiB` on HEAD of `teleterm` (5f1b0e7d7)

So Teleterm code adds 527.297KiB to the size of `tsh`. Built with `make build/tsh`.

(Also TIL that in Go usually there's no "release" build other than striping debug symbols.)

---

> It just seemed like an interesting design choice to ship extra code to _all users_ in tsh that is only needed for the terminal use case.

Yeah, I'm not sure if long term this is a good strategy either.

We didn't discuss an update strategy, the preview release won't have an update feature anyway and the RFD says that "`tsh` is packaged together with the rest of the application and installed into the "app data" folder". I'll ask the rest of the team if they made any plans in that area or if Teleterm will be tied to the general release schedule.

If Teleterm is always going to be bundled with `tsh`, then it would make sense to put the daemon code behind a build tag. If it's not going to pose any additional problems during development, then it's an easy way to reduce the attack surface and to just avoid shipping code that's not going to be used anyway.